internal/engine/compiler/engine.go

```diff
@@ -745,11 +754,13 @@ func (ce *callEngine) Call(ctx context.Context, params ...uint64) (results []uin
 	// This returns a safe copy of the results, instead of a slice view. If we
 	// returned a re-slice, the caller could accidentally or purposefully
 	// corrupt the stack of subsequent calls
-	if resultCount := tp.ResultNumInUint64; resultCount > 0 {
-		results = make([]uint64, resultCount)
-		copy(results, ce.stack[:resultCount])
+	if tp.ResultNumInUint64 > cap(results) {
+		results = make([]uint64, tp.ResultNumInUint64)
+	} else {
+		results = results[:tp.ResultNumInUint64]
 	}
-	return
+	copy(results, ce.stack[:tp.ResultNumInUint64])
+	return results, nil
```

Comment:
There's any specific reason to overwrite the params instead of "concatenate" it.

```go
	if tp.ParamNumInUint64 + tp.ResultNumInUint64 > cap(results) {
		results = make([]uint64, tp.ResultNumInUint64)
	} else {
		results = results[tp.ParamNumInUint64:tp.ParamNumInUint64 + tp.ResultNumInUint64]
	}
	copy(results, ce.stack[:tp.ResultNumInUint64])
	return results, nil
```


I think that will allow to have some function like:

```
param := make([]uint64, 2, 3)
param[0] = 42
param[1] = 7


for {
   res, err := fn.Call(ctx, param)
   // ...
}
```

In that case ` []uint64{42, 7}` will not be overwrite, and can be re-used across multiple calls. The response will be stored into the param[2], which is already allocated.

Reply:
No specific reason.

Basically, that's what a single argument stack API implies IMO: pop the arguments push the results. At least that's my interpretation of @codefromthecrypt's suggested API.

You and @achille-roussel seem to be converging on your `CallTo` suggested API: arguments are independent from results.

I honestly don't know which is better.

I'd like to see more usage scenarios to understand the advantages/disadvantages of either design.

Do you have examples you can share of how you'd use either API?

Reply:
I can't share the actual code, yet. But, I currently use an "Update" function that is called multiple times per second per instance. This function takes in two parameters, `id` and `offset`, and returns a uint32 value. The id is unique per instance, while the offset is rarely updated. So, I could re-use `params` between multiple calls. 

In that case, I would need to do something like: params[0] = id and params[1] = offset for each call.

I think `CallWithStack(ctx context.Context, params, results []uint64)` allows both worlds to exists:

If I would like to reduce some memory consumption and overwrite I use: `CallWithStack(ctx, params, params[:0])`
If I would like to re-use the params for future calls then I can use:  `CallWithStack(ctx, params, results)`

_In both cases, assuming that params/results have enough capacity._